bench/Bench/Utils.hs

```diff
@@ -0,0 +1,176 @@
+  => Proxy k
+  -> Proxy v
+  -> Limits
+     -- ^ LMDB limits such as maximum database size.
+  -> (Database k v -> Transaction ReadWrite ())
+     -- ^ Action that populate the database
+  -> m (BenchEnv k v)
+initBenchEnv _ _ limits populate = do
+  dbFilePath <- createDbDir
+  dbEnv <- liftIO $ openEnvironment dbFilePath limits
+  db <- liftIO $ readWriteTransaction dbEnv (getDatabase Nothing)
+  void $ liftIO $ readWriteTransaction dbEnv $ populate db
+  pure BenchEnv {dbFilePath, dbEnv, db}
+
+createDbDir :: MonadIO m => m FilePath
+createDbDir = do
+  sysTmpDir <- liftIO getTemporaryDirectory
+  liftIO $ createTempDirectory sysTmpDir "bench_cursor"
+
+cleanupBenchEnv :: MonadIO m => BenchEnv k v -> m ()
+cleanupBenchEnv tenv = do
+  liftIO $ closeEnvironmentBlocking (dbEnv tenv)
+  deleteDbDir (dbFilePath tenv)
+
+deleteDbDir :: MonadIO m => FilePath -> m ()
+deleteDbDir = liftIO . removeDirectoryRecursive
+
+{-------------------------------------------------------------------------------
+  Defaults for LMDB limits
+-------------------------------------------------------------------------------}
+
+simpleLMDBLimits :: Limits
+simpleLMDBLimits = defaultLimits
+
+{-------------------------------------------------------------------------------
+  Defaults for database population
+-------------------------------------------------------------------------------}
+
+-- | @'noPopulateDb' db@ does not populate the LMDB database @db@ with any
+-- entries.
+noPopulateDb :: Database k v -> Transaction ReadWrite ()
+noPopulateDb _ = pure ()
+
+-- | @'populateDbUnique' n db@ populates the LMDB database @db@ with @n@ unique
+-- key-value pairs.
+--
+-- Note: ensure that @n@ is much smaller than the range of values that are
+-- generated for @'Arbitrary' k@, or this function may not terminate within
+-- reasonable time, let alone terminate at all. We use a simple "retry if a
+-- generated key is already present in the database (i.e., a /clash/)" strategy
+-- to generate exactly @n@ unique keys, which means that clashes occur
+-- inhibitively often as @n@ approaches the size of the of the range of @k@.
+populateDbUnique ::
+     ( Serialise k, Serialise v
+     , Arbitrary k, Arbitrary v
+     , Ord k
+     )
+  => Int
+  -> Database k v
+  -> Transaction ReadWrite ()
+populateDbUnique n db
+  | n <= 0    = pure ()
+  | otherwise = do
+      k <- liftIO $ generate arbitrary
+      v <- liftIO $ generate arbitrary
+      b <- Internal.putNoOverwrite db k v
+      if b then
+        populateDbUnique (n - 1) db
+      else
+        populateDbUnique n db
+
+{-------------------------------------------------------------------------------
+  Default types for keys and values
+-------------------------------------------------------------------------------}
+
+newtype Key = Key B.ShortByteString
+  deriving stock (Show, Eq, Ord, Generic)
+  deriving newtype (NFData, Serialise)
+
+-- | Size of a @'Key'@ in number of bytes.
+keySize :: Int
```

Comment:
Why not have keys be `Word32` then? And why did we choose `32` bits?

Reply:
I simulate 32-byte and 64-byte cryptographic hashes like I did in https://github.com/input-output-hk/ouroboros-network/pull/4143, `Word32`s would be too small (they are 32-bit, not 32-byte)